File: Algorithms/QuickSort/QuickSort.ts
export default function quickSort(arr: number[]): void {
    qs(arr, 0, arr.length - 1);
}

function qs(arr: number[], low: number, high: number): void {
    if (low >= high)
        return;

    const pivotIdx = partition(arr, low, high);
    qs(arr, low, pivotIdx - 1);
    qs(arr, pivotIdx + 1, high);
}

function partition(arr: number[], low: number, high: number): number {
    const pivot = arr[high];

    let idx = low - 1;

    for (let i = low; i < high; i++) {
        if (arr[i] <= pivot) {
            idx++;
            const tmp = arr[i];
            arr[i] = arr[idx];
            arr[idx] = tmp;
        }
    }

    idx++;
    arr[high] = arr[idx];
    arr[idx] = pivot;

    return idx;
}

File: C/src/algorithms/quick-sort/quick-sort.c
#include "quick-sort.h"

void quickSortRecurse(int arr[], int l, int h);
int partition(int arr[], int l, int h);

void quickSort(int arr[], int n)
{
    int l = 0;
    int h = n - 1;

    quickSortRecurse(arr, l, h);
}

void quickSortRecurse(int arr[], int l, int h)
{
    if (l >= h)
        return;

    int pIndex = partition(arr, l, h);

    quickSortRecurse(arr, l, pIndex - 1);
    quickSortRecurse(arr, pIndex + 1, h);
}

int partition(int arr[], int l, int h)
{
    int pivot = arr[h];
    int i = l;

    for (int j = l; j < h; j++)
    {
        if (arr[j] <= pivot)
        {
            int tmp = arr[j];
            arr[j] = arr[i];
            arr[i] = tmp;
            i++;
        }
    }

    int tmp = arr[h];
    arr[h] = arr[i];
    arr[i] = tmp;

    return i;
}